### DISCUSSION TITLE
hPyT v1.3.0

### DISCUSSION THREAD

**Discussion by Zingzy** [Zingzy/hPyT#13] [Announcements]
### v1.3.0

- Add support for setting custom border color
- Add support for rainbow border color effect
- Add support for resetting the titleBar color and titleText color
- Fix an issue which caused the titleBar to appear black after the rainbow titleBar effect was stopped

For detailed information and previews please refer the README.md file

<hr /><em>This discussion was created from the release <a href='https://github.com/Zingzy/hPyT/releases/tag/v1.3.0'>hPyT v1.3.0</a>.</em>

**Comment by Sae3sy** [comment]
could you also please add a function to get the current color used by hPyT so you could synchronize other elements with it?

**Reply by Zingzy** [reply]
Hello @Sae3sy can you please elaborate about the feature you are requesting?

**Reply by Sae3sy** [reply]
simply a function that returns the current color value of rainbow titlebar or border. With such a function you could synchronize other interface elements e. g. a progress bar that is changing color with the title bar

**Reply by Sae3sy** [reply]
would you do it?

**Reply by Zingzy** [reply]
Yeah, I would try to add it

**Reply by Sae3sy** [reply]
would also be fantastic if you give me a update as you progress on it :)) Thanks!

**Reply by Zingzy** [reply]
Hey, I am extremely sorry, I have been a lot busy for the past 2 weeks and I would try my best to implement this feature by the next week.

**Reply by Sae3sy** [reply]
No problem I can wait, it's not a necessary feature but it would be cool to have. And I have the code so I could do it myself - although I'm also busy and I would need sooome time to get familiar with the code because by now I don't really understand how it works. Thank you!

**Reply by Zingzy** [reply]
Hello @Sae3sy a quick update, I was thinking about the requested feature and all I could come up with was adding something like this:
```py
class rainbow_title_bar:
    current_color = None  # Class variable to store the current color

    @classmethod
    def start(cls, window, interval=5, color_stops=5) -> None:
        def color_changer(hwnd, interval):
            r, g, b = 200, 0, 0
            while hwnd in rnbtbs:
                cls.current_color = (r << 16) | (g << 8) | b 
                ...

    @classmethod
    def get_current_color(cls):
        return cls.current_color
```
this can be called this way:
```py
rainbow_title_bar.get_current_color() #this will return a tuple containing the current color
```

were you expecting something like this? If you have better ideas/plans for this feature please feel free to discuss!
Again thanks for showing so much support and patience 💛

**Comment by Sae3sy** [comment]
cool it seems to work! Though I maybe too stupid to use it. So like I said I wanted to create a progressbar which changes color in sync with your rainbow title bar. So I created this simple code for testing purposses:

```python
import sys
from PyQt6.QtWidgets import QApplication, QMainWindow, QProgressBar, QVBoxLayout, QWidget
from PyQt6.QtCore import QTimer
import qdarktheme
from hPyT import rainbow_title_bar

class RainbowProgressBar(QProgressBar):
    def __init__(self, *args, interval=100, color_stops=5, **kwargs):
        super().__init__(*args, **kwargs)
        self.setTextVisible(False)
        self.interval = interval
        self.color_stops = color_stops
        self.timer = QTimer()
        self.timer.timeout.connect(self.update_color)
        self.timer.start(interval)
        self.setStyleSheet("QProgressBar::chunk { background-color: rgb(255, 0, 0); }")

    def update_color(self):
        color = rainbow_title_bar.get_current_color()
        r = (color >> 16) & 0xFF
        g = (color >> 8) & 0xFF
        b = color & 0xFF
        self.setStyleSheet(f"QProgressBar::chunk {{ background-color: rgb({r}, {g}, {b}); }}")

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Rainbow ProgressBar hPyT sync")
        self.setGeometry(100, 100, 600, 400)
        
        qdarktheme.setup_theme("dark")

        self.progress = RainbowProgressBar(self)
        self.progress.setGeometry(50, 50, 500, 50)
        self.progress.setRange(0, 100)
        self.progress.setValue(50)  # Example progress

        layout = QVBoxLayout()
        layout.addWidget(self.progress)
        container = QWidget()
        container.setLayout(layout)
        self.setCentralWidget(container)

if __name__ == "__main__":
    app = QApplication(sys.argv)

    # Start the rainbow title bar effect for the window
    window = MainWindow()
    hwnd = int(window.winId())
    rainbow_title_bar.start(window, interval=1, color_stops=5)  # Example interval and stops

    window.show()
    sys.exit(app.exec())
```

The problem is that even if I set the interval to the same value the colors aren't the same - so while searching for a solution I found a weird bug that I think could be the source of it why it isn't working: changing the interval of the ProgressBar class does influence how fast is the color of the title bar changing.

https://github.com/Zingzy/hPyT/assets/69053338/b93acd2d-97f9-453c-aeb0-48ccdd37b4bb

**Reply by Zingzy** [reply]
Hi, did my method work for you? Also your method seems fine and should have worked, there might be some bugs or delays which might lead to unsynchronization.
About the bug that you mentioned that changing the interval doesnot change the speed, I tried to test it and wierdly it works just fine for me.

https://github.com/Zingzy/hPyT/assets/90309290/50e0faeb-ad72-4fbe-a232-3ffd52b72d40

**Reply by Sae3sy** [reply]
changing the interval works, but when I change the interval of the progressbar not the titlebar it does also influence how fast is the titlebar changing colors. Notice that in the video I posted in the comment, I only changed the progressbar interval and the titlebar also slowed down, which shouldn't happen.

**Reply by Zingzy** [reply]
@Sae3sy Good news! I seemed to have identified the reason why the colors where not synchronizing properly... you see the color in the `rainbow_title_bar` method is initially a `hex notation` which has the `r` and `g` values swapped (more clearly explained in this webpage: https://spoo.me/hexNotation) 

I have modified the code to return the colors correctly:

**`rainbow_title_bar` code**:
```py
@classmethod
    def get_current_color(cls):
        color = cls.current_color
        b = (color >> 16) & 0xFF
        g = (color >> 8) & 0xFF
        r = color & 0xFF
        return (r, g, b)
```
**You code**:
```py
def update_color(self):
        r,g,b = rainbow_title_bar.get_current_color()
        self.setStyleSheet(f"QProgressBar::chunk {{ background-color: rgb({r}, {g}, {b}); }}")
```

This fixes the asynchronization error. 

https://github.com/Zingzy/hPyT/assets/90309290/f75d6850-54ba-4b4c-aa87-722903fbcb6e

About the interval mixing issue, I am still trying to identify the bug.

**Reply by Sae3sy** [reply]
you're the best! Thank you!

**Reply by Zingzy** [reply]
no worries, always happy to help 😀
Also, I would love it if you could **contribute the final working version of this app** to hPyT/Examples folder as `synchronization` example?
I would release this feature in the main library very soon

**Reply by Sae3sy** [reply]
yeah, no problem, I would love to do that! I will create a more precised and better looking example in the next days because I am very busy now.

**Reply by Zingzy** [reply]
Ok please take your time, I am not in a hurry to make a new release 😄
